src/codesyncd/validators/tab_validator.ts

```diff
@@ -0,0 +1,63 @@
+/*
+class TabValidator:
+	validateRepoID()
+	validate_yml_file()
+	group_data()
+*/
+
+import { REQUIRED_KEYS_TAB_FILE_YML, TAB_SIZE_LIMIT } from "../../constants";
+import { ITabYML } from "../../interface";
+import { CodeSyncLogger } from "../../logger";
+import { generateSettings } from "../../settings";
+import { readYML } from "../../utils/common";
+import { ConfigUtils } from "../../utils/config_utils";
+import { removeFile } from "../../utils/file_utils";
+
+export class TabValidator {
+
+	settings: any;
+	configJSON: any;
+	constructor() {
+		this.settings = generateSettings();
+        this.configJSON = readYML(this.settings.CONFIG_PATH);
+
+	}
+
+	validateRepo(tabFile: ITabYML, tabData: string, filePath: string, repoId: number) {
+		const config_utils = new ConfigUtils();
+		const repo_path = config_utils.getRepoPathByRepoId(repoId);
+		const configRepo = this.configJSON.repos[repo_path];
+		if (!configRepo) {
+			CodeSyncLogger.info(`Removing tab: Skipping invalid tab: ${tabFile}`, "", tabData);
+			removeFile(filePath, "getTabFiles");
+			return false;
+		}
+		// Remove tab if repo is disconnected
+		if (configRepo.is_disconnected) {
+			CodeSyncLogger.error(`Removing tab: Repo ${repo_path} is disconnected`);
+			removeFile(filePath, "getTabFiles");
+			return false;
+		}
+		return true;
+	}
+	
+	validateRepoId(tabData: ITabYML, repo_id: number) {
+		console.log(`tabData.repo_id: ${tabData.repository_id}, repo_id: ${repo_id}`);
+		if(tabData.repository_id === repo_id) {
```

Comment:
Didn't get this validation, can you explain?

Reply:
I have updated this a bit, I was passing `tabData.repo_id` instead of the `repo_id` that we were computing.

Let me give you an overview of what we are doing:

We are retrieving the path of tab and removing the last segment, so that it is equal to current_repo:

![image](https://github.com/codesyncapp/codesync-vs/assets/15974895/d4c9fb4d-c4dd-423e-b5f9-1198dd81c681)

e.g. `home/documents/saad/test-project` would become `home/documents/saad/`

We then use config util: `getRepoIdByPath` to get repo_id.

After that, we pass the current tab data object (which has keys `repository_id`, `source` etc), and the repo_id we computed.:

![image](https://github.com/codesyncapp/codesync-vs/assets/15974895/7c894cbc-e0a5-407e-a356-d7ea867f7982)

And in our validator, we ensure that the repo_id computed from tab path, and the original repo_id is the same.

Reply:
Alright, I don't think we need "data" validation, server will raise 403 if user tries to play with `.tabs` directory, since we are adding data to buffer, we can assume data is fine, however validating structure is fine

Reply:
If I understood this correctly, What if the TabFile is in a nested directory and not on the root path?
e.g. 
```home/documents/saad/test-project/dir/dir2/abc.ts```
would become
```home/documents/saad/test-project/dir/dir2```

Which is NOT the repository path and we will not get any repoID against it

> We are retrieving the path of tab and removing the last segment, so that it is equal to current_repo

Reply:
@saadberry Suggested solution:

```
// repoPath = "/home/documents/saad/test-project"
// tabfilePath1 = /home/documents/saad/test-project/dir/dir2/abc.ts
// tabfilePath2 = "/home/documents/saad/test-project-2/dir/file.ts"
// const splitPath = tabfilePath1.split(`${repoPath}${path.sep}`);
// splitPath.length == 2
// const splitPath2 = tabfilePath2.split(`${repoPath}${path.sep}`);
// splitPath2.length == 1

If splitPath.length ==0; ignore that tab, otherwise pick index = 1
```